Add options to modify/disable certain dependencies

Zaxus125 opened the pull request:
This PR has been split up into the following components:
 - `NO_PULSE` - #1922
 - `USE_OPENGLES` - #1924
 - `NO_GLEW` - #1925
 - `USE_X11` - TBD
 - `BUILD_OTR_GUI` - No longer needed
---

## Motivation
I would like to see SoH able to be run on small embedded devices, such as a Raspberry Pi, or my goal in particular, retro handhelds such as the Anbernic RG series. This PR is a step in that direction, by allowing certain dependencies to be trimmed/replaced with alternatives that are better suited for embedded Linux environments.

This PR is marked as a draft because there is still some cleaning up that needs to be done, but I want to get feedback first to determine if it is worth spending more time on it.

## Changes
The following options to libultraship's CMake have been added
 - `NO_GLEW` - Set to `ON` to use SDL's extensions instead of GLEW.
 - `USE_OPENGLES` - Set to `ON` to link against `libGLESv2` instead of `libGL`. This requires using the newer glsl that MacOS uses.
 - `NO_PULSE` - Set to `ON` to use SDL Audio instead of the pulseaudio backend.
 - `USE_X11` - When set to `OFF`, do not link against `-lX11`. The default is now `ON` (currently, nothing defines `X11_SUPPORTED`)
 - `BUILD_OTR_GUI` - Set to `ON` to build the `OTRGui` tool. Defaults to `ON`. Will still not build on systems matching `Darwin|NintendoSwitch|CafeOS`. 

## Building
This can be tested by supplying the following options during the CMake generation step:
```bash
$ cmake -H. -Bbuild-cmake -GNinja -DNO_GLEW=ON -DNO_PULSE=ON -DUSE_OPENGLES=ON -DUSE_X11=OFF -DBUILD_OTR_GUI=OFF

# other commands as normal
$ cmake --build build-cmake --target ExtractAssets
$ cmake --build build-cmake
```

## Known Issues / Open Questions / Misc Implementation Details
 - `pkg-config --libs glew` includes `-lX11` as one of the flags, so `X11=0` might also depend on `NO_GLEW=1`
 - It seems the naming convention for Makefile options is the opposite of what I implemented (e.g. `X11=0` vs `NO_X11=1`), will go back and change if needed
 - The fragment shader uses `#version 130`, because it needs `textureSize`. Unfortunately, the earliest ES version that supplies that function is `#version 300 es`, which is based on glsl version 3.30. Since MacOS already has a newer glsl version, this borrows that implementation. I honestly don't know what the implications are of this, I am not terribly experienced with OpenGL.
 -  ~~`ZAPDTR` links against `libultraship.a`, which pulls in a lot of the same linker dependencies as `soh.elf`. I suspect a lot of those can also be wholesale trimmed if the entire archive isn't linked in, and only certain files are included instead. This would also aid cross compilation builds, but that seems out of scope for this PR.~~ This may have been addressed by the CMake overhaul

## Testing
So far, I have only tested these changes on my Linux desktop. Unless I messed up an `#ifdef` somewhere, this should not affect Windows/Mac/Switch, but I will look into personally testing on at least Windows.

I am aware that adding a lot of options introduces a lot of complexity with how many combinations of libraries are used. I have tried to make sure that each current option does not affect others (exception being that building with GLEW implies `-lX11`), and have at least compile tested all default, all enabled, and each individual options.

## Results
The following is the output of `ldd soh.elf` on my system before and after these changes:
```
# Regular make with no options set
$ ldd soh.elf 
	linux-vdso.so.1 (0x00007ffe247ef000)
	libSDL2-2.0.so.0 => /usr/lib/libSDL2-2.0.so.0 (0x00007eff6a24f000)
	libbz2.so.1.0 => /usr/lib/libbz2.so.1.0 (0x00007eff6b37b000)
	libz.so.1 => /usr/lib/libz.so.1 (0x00007eff6b361000)
	libGLEW.so.2.2 => /usr/lib/libGLEW.so.2.2 (0x00007eff6a192000)
	libGL.so.1 => /usr/lib/libGL.so.1 (0x00007eff6b2db000)
	libX11.so.6 => /usr/lib/libX11.so.6 (0x00007eff6a04f000)
	libGLU.so.1 => /usr/lib/libGLU.so.1 (0x00007eff69ff9000)
	libOpenGL.so.0 => /usr/lib/libOpenGL.so.0 (0x00007eff69fcd000)
	libpulse.so.0 => /usr/lib/libpulse.so.0 (0x00007eff69f78000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007eff69c00000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007eff69e91000)
	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007eff69e71000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007eff69800000)
	/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007eff6b3da000)
	libGLdispatch.so.0 => /usr/lib/libGLdispatch.so.0 (0x00007eff69b48000)
	libGLX.so.0 => /usr/lib/libGLX.so.0 (0x00007eff69e3e000)
	libxcb.so.1 => /usr/lib/libxcb.so.1 (0x00007eff69b1d000)
	libpulsecommon-16.1.so => /usr/lib/pulseaudio/libpulsecommon-16.1.so (0x00007eff69a95000)
	libdbus-1.so.3 => /usr/lib/libdbus-1.so.3 (0x00007eff69a42000)
	libXau.so.6 => /usr/lib/libXau.so.6 (0x00007eff6b2d0000)
	libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0x00007eff6b2c8000)
	libsndfile.so.1 => /usr/lib/libsndfile.so.1 (0x00007eff69780000)
	libsystemd.so.0 => /usr/lib/libsystemd.so.0 (0x00007eff696a2000)
	libasyncns.so.0 => /usr/lib/libasyncns.so.0 (0x00007eff69a3a000)
	libvorbisenc.so.2 => /usr/lib/libvorbisenc.so.2 (0x00007eff695f7000)
	libFLAC.so.8 => /usr/lib/libFLAC.so.8 (0x00007eff695b9000)
	libopus.so.0 => /usr/lib/libopus.so.0 (0x00007eff6955f000)
	libvorbis.so.0 => /usr/lib/libvorbis.so.0 (0x00007eff69a0c000)
	libogg.so.0 => /usr/lib/libogg.so.0 (0x00007eff69554000)
	libcap.so.2 => /usr/lib/libcap.so.2 (0x00007eff69548000)
	libgcrypt.so.20 => /usr/lib/libgcrypt.so.20 (0x00007eff693ff000)
	liblzma.so.5 => /usr/lib/liblzma.so.5 (0x00007eff693d5000)
	libzstd.so.1 => /usr/lib/libzstd.so.1 (0x00007eff6932c000)
	liblz4.so.1 => /usr/lib/liblz4.so.1 (0x00007eff69309000)
	libgpg-error.so.0 => /usr/lib/libgpg-error.so.0 (0x00007eff692e3000)
```
```
# make NO_GLEW=1 NO_PULSE=1 USE_OPENGLES=1
$ ldd soh.elf
	linux-vdso.so.1 (0x00007ffcc9d75000)
	libSDL2-2.0.so.0 => /usr/lib/libSDL2-2.0.so.0 (0x00007f211ea4f000)
	libbz2.so.1.0 => /usr/lib/libbz2.so.1.0 (0x00007f211fc4b000)
	libz.so.1 => /usr/lib/libz.so.1 (0x00007f211fc31000)
	libGLESv2.so.2 => /usr/lib/libGLESv2.so.2 (0x00007f211fc1e000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007f211e800000)
	libm.so.6 => /usr/lib/libm.so.6 (0x00007f211fb37000)
	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007f211fb15000)
	libc.so.6 => /usr/lib/libc.so.6 (0x00007f211e400000)
	/lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007f211fcaa000)
	libGLdispatch.so.0 => /usr/lib/libGLdispatch.so.0 (0x00007f211e748000)
```

Zaxus125 committed:
add OpenGLES support

 - Use glsl #version 300 es
 - Change the #if guards on glsl generation to USE_NEWER_OPENGL
   - use this for both USE_OPENGLES and __APPLE__ build targets
 - Tweak glsl for ES-specific issue with implicit ivec -> vec conversion
 - Have ImGui figure out its own glsl version

Zaxus125 committed:
add NO_GLEW option to build using SDL extensions over GLEW

Zaxus125 committed:
add NO_PULSE to build using only SDL audio

vanfanel commented:
Yes, please. Definitely yes. As I said, I would sponsor this if needed.

The very meaning of SOH is, for me, to get it truly multiplatform, and as of today that includes ARM devices like handhelds like the RG series and the worldwide ultra-popular computer of the masses: the Raspberry Pi.
Building on the Pi was already somewhat possible, but performance on GLX is atrocious (not to mention input lag...) and it has inherent X11 tearing which destroys the game's visuals.
Using the right abstraction API (SDL2) instead of going for GLX directly is the way to go: SDL2 in turn can make SOH run on the old X11 (GLX), KMS/DRM (no graphics server, ultra low-latency option) or even Wayland.
Same goes for audio.

Particularly, the operating systems on those GNU/Linux handhelds (OGA, RG series, and a long etc) lack GLX for good, so ports ONLY target SDL2 which in turn provide 3D support using OpenGL, OpenGL_ES or even Vulkan. (Yes, Vulkan is possible without X11: I implemented a solution so a Vulkan context is possible on SDL2 without a graphics server).

As an example, SM64 ports are everywhere: and that's because it targets the right abstraction API (SDL2) instead of pointing to specific OpenGL implementations.
The same can be said about other very popular engines like Scummvm, ChocolateDoom, SDLPop, RigelEngine, and a long, long list of engines that build and play perfectly everywhere because they target SDL2 instead of pointing to playtform-specific APIs.

Also, this is truly beautiful to see how dependencies get WAY smaller after this. Truly beautiful to see that, and the immense implications that has for portability.

So I will build ASAP and report. This is a dream come true. Thanks for making it start rolling!

awe444 commented:
@Zaxus125 I’ve tested these changes in a Raspberry Pi 4 build and I’m pleased to say they work wonderfully. Finally can run the game on the Pi without Mesa compatibility mode and without pulseaudio!

I used native built Mesa and wayland as the compositor for testing. Unfortunately didn’t see much performance improvement relative to the previous setup which used compatibility mode.

keep up the great work!

Zaxus125 committed:
WIP: add cmake options for NO_GLEW, NO_PULSE, USE_OPENGLES, USE_X11

Zaxus125 commented:
Updated to now use the new CMake build system. Original PR message has been updated with new build instructions.

Some additional notes:
 - It appears currently nothing defines `X11_SUPPORTED` that was added when the switch port was merged. Therefore, current upstream has no way of enabling glx support, yet still links against it. The `USE_X11` option restores the ability to use glx, and enables compiling it by default.
 - The commit to CMakeLists is very much a WIP:
   - The option names are as always subject to change, feel free to suggest better/more consistent names
   - The placement of option/dependency setting logic can probably be simplified and/or relocated, not sure where is best though

Zaxus125 committed:
add option to build the OTRGui tool, defaults to ON

Zaxus125 commented:
Added an option to disable building `OTRGui`, which still depends on X11. As the commit message says, this method is a hack and should be implemented better -- perhaps as a target, or a dependency to a target like release/packaging. Open to suggestions on how to improve that.

Zaxus125 committed:
add -DNO_PULSE if NO_PULSE=ON when building soh

OTRGlobals.cpp includes AudioPlayer.h from libultraship, which pulls in
PulseAudio headers. This points to the build options probably belonging in the
root CMakeLists rather than in libultraship's. It works here, but it should
definitely be moved to a more common location.

This works around building soh via a cross-compilation toolchain and/or on a
system that does not have PulseAudio.

awe444 commented:
OK, so I managed to build and run successfully on my Pi 4 from the latest cmake-based commits in this PR. Once again, great work @Zaxus125 !

~However, I did have to locally revert the merge commit [6818247](https://github.com/HarbourMasters/Shipwright/commit/68182473176510de5b2d28b6bc3a4a4fc9e0b022) from PR https://github.com/HarbourMasters/Shipwright/pull/909 because there's something in that commit that broke the Pi builds-- it _might_ have something to do with the fact that I'm using **32-bit** Pi OS, but I haven't tried building on 64-bit Pi OS so that's just speculating. Without reverting that commit I get a binary which segfaults upon launching~  **EDIT**: per my newer comment below, this issue was resolved by migrating to 64-bit OS.

I also migrated (back) to a totally X11-free, pulseaudio-free setup in order to perform this testing, since that was kind of the point of these changes in the first place. Here's my setup details (again note that it's 32-bit OS):

- Linux raspberrypi 5.15.32-v7l+ #1538 SMP Thu Mar 31 19:39:41 BST 2022 armv7l GNU/Linux
- cmake version 3.24.20220808-g33d2491
- GL version: OpenGL ES 3.1 Mesa 22.3.0-devel (git-dccd6f495a)
- GLSL version: OpenGL ES GLSL ES 3.10
- SDL 2.23.1
- GLEW which was built with `SYSTEM=linux-egl-glvnd` 
- Launched SoH with weston 10.0.90

My cmake command was as follows. I left in GLEW as a dependency since I have it around for other applications and figured why not.
```
export CFLAGS="-O3 -march=armv8-a+crc -mtune=cortex-a72"
export CPPFLAGS="$CFLAGS"
export CXXFLAGS="$CFLAGS"
cmake -H. -Bbuild-cmake -GNinja \
-DNO_PULSE=ON \
-DUSE_OPENGLES=ON \
-DUSE_X11=OFF \
-DBUILD_OTR_GUI=OFF \
-DCMAKE_BUILD_TYPE:STRING=Release
```
Library linkages:
```
 ldd ./soh.elf
        linux-vdso.so.1 (0xbef68000)
        /usr/lib/arm-linux-gnueabihf/libarmmem-${PLATFORM}.so => /usr/lib/arm-linux-gnueabihf/libarmmem-v7l.so (0xb6efc000)
        libSDL2-2.0.so.0 => /lib/libSDL2-2.0.so.0 (0xb6d92000)
        libGLEW.so.2.2 => /lib/libGLEW.so.2.2 (0xb6cdf000)
        libGLESv2.so.2 => /lib/arm-linux-gnueabihf/libGLESv2.so.2 (0xb6cb9000)
        libz.so.1 => /lib/arm-linux-gnueabihf/libz.so.1 (0xb6c91000)
        libbz2.so.1.0 => /lib/arm-linux-gnueabihf/libbz2.so.1.0 (0xb6c71000)
        libstdc++.so.6 => /lib/arm-linux-gnueabihf/libstdc++.so.6 (0xb6ae9000)
        libm.so.6 => /lib/arm-linux-gnueabihf/libm.so.6 (0xb6a7a000)
        libgcc_s.so.1 => /lib/arm-linux-gnueabihf/libgcc_s.so.1 (0xb6a4d000)
        libpthread.so.0 => /lib/arm-linux-gnueabihf/libpthread.so.0 (0xb6a21000)
        libc.so.6 => /lib/arm-linux-gnueabihf/libc.so.6 (0xb68cd000)
        libdl.so.2 => /lib/arm-linux-gnueabihf/libdl.so.2 (0xb68b9000)
        libEGL.so.1 => /lib/arm-linux-gnueabihf/libEGL.so.1 (0xb6897000)
        libGLdispatch.so.0 => /lib/arm-linux-gnueabihf/libGLdispatch.so.0 (0xb675f000)
        /lib/ld-linux-armhf.so.3 (0xb6f11000)
```

Update: I tried building on 64-bit Pi OS and unfortunately the compiling runs out of memory on my meager 2GB Pi 4, even with `-j 1` flag. If anyone else has built on 64-bit Pi OS please share your findings.

foxhound311 commented:
i also got it up and running on my 2gb rpi4. fresh install of 32bit rpiOS bullseye.
used same method as @awe444, to be safe i went with mesa 20.3.5 and did this instructions: 
$ git clone --branch less-dependencies-pr https://github.com/Zaxus125/Shipwright
$ cd Shipwright
$ git revert 68182473176510de5b2d28b6bc3a4a4fc9e0b022 -X ours --no-edit
$ sed libultraship/libultraship/color.h -i -e 's|#include "endianness.h"||g'
$ export CFLAGS="-O3 -march=armv8-a+crc -mtune=cortex-a72"
export CPPFLAGS="$CFLAGS"
export CXXFLAGS="$CFLAGS"
cmake -H. -Bbuild-cmake -GNinja \
-DNO_PULSE=ON \
-DUSE_OPENGLES=ON \
-DUSE_X11=OFF \
-DBUILD_OTR_GUI=OFF \
-DCMAKE_BUILD_TYPE:STRING=Release
$ cmake --build build-cmake --target ExtractAssets -j2
$ cmake --build build-cmake -j2

took about 70 minutes to compile but im sure -j4 can also be done.
game runs perfectly fine. still need to play more to find bugs and such.  

all credit goes to @Zaxus125  and @awe444, great work.

retropieuser commented:
This looks pretty interesting. Is there any options to have this work on CMake 3.16? I attempted this on Retropie and the cmake won't run as it needs CMake 3.19 and above. But retropie being still on the Buster rather than Bulleyes RPI OS Lite, it doesn't seem happy to upgrade above 3.16?

Anyway I'll have a try on a fresh install of 32bit rpiOS bullseye. Great work!

awe444 commented:
> I attempted this on Retropie and the cmake won't run as it needs CMake 3.19 and above

It was actually buildable on the Pi even before this PR and before the cmake integration, i.e., if you go back a sufficient number of commits you could build it with the Makefile. Not sure it’d be worth it though since you’ll need apply hacks for ARM compatibility (which have since been integrated into the main branch) and you’d need to use Mesa compatibility mode.

Alternatively if you want to install latest cmake from source, try:
```
git clone https://github.com/Kitware/CMake
cd CMake
./bootstrap --prefix=/usr
make -j 4
make install
```

Zaxus125 commented:
Rebased onto current upstream, removed the `BUILD_OTR_GUI` option since 791f777 appears to take care of that.

> Is there any options to have this work on CMake 3.16

~I just pushed some experimental work on this to another branch here: https://github.com/Zaxus125/Shipwright/tree/extract-assets-arg . That branch only contains that single commit, so you might want to cherry pick that commit onto this PR branch to test both.~

~Disclaimer: I have barely tested that change, will probably open a separate pull request for that once I confirm it works and doesn't break something else.~

Opened a PR to lower CMake version to 3.16, to test on this PR, cherrypick commit d7359ff from my [`lower-required-cmake` branch](https://github.com/Zaxus125/Shipwright/tree/lower-required-cmake).

awe444 commented:
Having some trouble building the newest commits (with the cherry-picked lower-cmake change). Here are my commands:
```
git clone --branch less-dependencies-pr https://github.com/Zaxus125/Shipwright
cd Shipwright
git cherry-pick d7359ff176fa816b7d02821c462ab881a2f4ec0a -X theirs --no-edit
cmake -H. -Bbuild-cmake -GNinja -DNO_PULSE=ON -DUSE_OPENGLES=ON -DUSE_X11=OFF -DCMAKE_BUILD_TYPE:STRING=Release
```
This errors during the cmake call as seen below:
```
Cloning into 'Shipwright'...
remote: Enumerating objects: 14686, done.
remote: Total 14686 (delta 0), reused 0 (delta 0), pack-reused 14686
Receiving objects: 100% (14686/14686), 63.09 MiB | 11.57 MiB/s, done.
Resolving deltas: 100% (8285/8285), done.
Updating files: 100% (5560/5560), done.
Removing CMake/genscript.cmake
[less-dependencies-pr 2103da9] supply ZAPD.out path as an argument to extract_assets.py rather than regenerating the script
 Author: Zaxus <zaxus125@proton.me>
 Date: Sat Aug 13 16:24:33 2022 -0500
 4 files changed, 16 insertions(+), 19 deletions(-)
 delete mode 100644 CMake/genscript.cmake
-- The C compiler identification is GNU 10.2.1
-- The CXX compiler identification is GNU 10.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Linux Raspbian 11 bullseye
-- Found ZLIB: /usr/lib/arm-linux-gnueabihf/libz.so (found version "1.2.11")
-- Found BZip2: /usr/lib/arm-linux-gnueabihf/libbz2.so (found version "1.0.8")
-- Looking for BZ2_bzCompressInit
-- Looking for BZ2_bzCompressInit - found
-- Linking against dependent libraries statically
-- Found GLEW: /usr/include (found version "2.2.0")
-- Disabling X11/GLX
-- Disabling Pulse
-- Building with OpenGLESv2
-- Found PNG: /usr/lib/arm-linux-gnueabihf/libpng.so (found version "1.6.37")
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE
-- Performing Test COMPILER_HAS_THOSE_TOGGLES
-- Performing Test COMPILER_HAS_THOSE_TOGGLES - Success
-- Testing if -Werror=pointer-arith can be used -- compiles
-- Testing if -Werror=implicit-function-declaration can be used -- compiles
-- Testing if -fno-strict-aliasing can be used -- compiles
-- Using raylib's GLFW
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Using X11 for window creation
CMake Error at /usr/share/cmake-3.24/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find X11 (missing: X11_X11_INCLUDE_PATH X11_X11_LIB)
Call Stack (most recent call first):
  /usr/share/cmake-3.24/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.24/Modules/FindX11.cmake:481 (find_package_handle_standard_args)
  OTRGui/libs/raylib/src/external/glfw/CMakeLists.txt:156 (find_package)


-- Configuring incomplete, errors occurred!
```

Zaxus125 commented:
Rebased onto current upstream again, includes the lowered CMake version now, so no need to cherry-pick.

Turns out I didn't read that commit I linked earlier closely enough, it still tries to source all the dependencies for OTRGui, even if you aren't building it. Will fix that when I have some more time, will most likely tie it to the `USE_X11` option, or skip it entirely if we don't have X11.

Here's a temporary fix if you want to work around it:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index f42fb69c..c3cf566f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -122,7 +122,7 @@ add_subdirectory(ZAPDTR/ZAPD ${CMAKE_BINARY_DIR}/ZAPD)
 add_subdirectory(ZAPDTR/ZAPDUtils ${CMAKE_BINARY_DIR}/ZAPDUtils)
 add_subdirectory(OTRExporter)
 add_subdirectory(soh)
-if(NOT CMAKE_SYSTEM_NAME MATCHES "Darwin|NintendoSwitch|CafeOS")
+if(NOT CMAKE_SYSTEM_NAME MATCHES "Darwin|NintendoSwitch|CafeOS|Linux")
 add_subdirectory(OTRGui)
 endif()
 ```
 
 **UPDATE**: Re-added the commit for now. Not sure yet where all the options should live, so not going to tangle all that logic up yet. Make sure to also set `BUILD_OTR_GUI=OFF` again.

Zaxus125 committed:
copy paste error, fix duplicate NO_GLEW with NO_PULSE

retropieuser commented:
So I tried this and it seems to get stuck on the cmake list file still on a pi4 with retropie:-

https://pastebin.com/q8MAdWTW

What am I doing wrong here? This is my cmake.text file:-

https://pastebin.com/fgwD38Jq

foxhound311 commented:
> So I tried this and it seems to get stuck on the cmake list file still on a pi4 with retropie:-
> 
> https://pastebin.com/q8MAdWTW
> 
> What am I doing wrong here? This is my cmake.text file:-
> 
> https://pastebin.com/fgwD38Jq

You need the dependencies, looks like you’re missing sdl2, glew and you prolly need mesa

retropieuser commented:
for libglew-dev I installed that and for sdl2 I think Retropie installs a version of sdl2 based on this:- https://github.com/RetroPie/RetroPie-Setup/blob/master/scriptmodules/supplementary/sdl2.sh Is that version not going to work?

awe444 commented:
I've finally managed to build and run this branch successfully on 64-bit Pi OS, without having to revert the endianness commit. Thanks a ton to @foxhound311 for the suggestion to increase swap file size to allow enough memory for compilation!

Here are the build commands that currently work for me on 64-bit Pi OS:
```
cmake -H. -Bbuild-cmake -GNinja -DNO_PULSE=ON -DUSE_OPENGLES=ON -DUSE_X11=OFF -DBUILD_OTR_GUI=OFF
cmake --build build-cmake --target ExtractAssets -j 3
cmake --build build-cmake --config Release -j 2
```
(again I'm still including GLEW rather than opting to omit it).

Other information on my setup:
```
cmake version 3.24.20220820-g9e884d4

Distributor ID: Debian
Description:    Debian GNU/Linux 11 (bullseye)
Release:        11
Codename:       bullseye

Linux raspberrypi4 5.15.56-v8+ #1575 SMP PREEMPT Fri Jul 22 20:31:26 BST 2022 aarch64 GNU/Linux

GL version: OpenGL ES 3.1 Mesa 22.3.0-devel (git-7b60414494)
GLSL version: OpenGL ES GLSL ES 3.10

weston 10.0.91
```

foxhound311 commented:
> Rebased onto current upstream again, includes the lowered CMake version now, so no need to cherry-pick.
> 
> Turns out I didn't read that commit I linked earlier closely enough, it still tries to source all the dependencies for OTRGui, even if you aren't building it. Will fix that when I have some more time, will most likely tie it to the `USE_X11` option, or skip it entirely if we don't have X11.
> 
> Here's a temporary fix if you want to work around it:
> 
> ```diff
> diff --git a/CMakeLists.txt b/CMakeLists.txt
> index f42fb69c..c3cf566f 100644
> --- a/CMakeLists.txt
> +++ b/CMakeLists.txt
> @@ -122,7 +122,7 @@ add_subdirectory(ZAPDTR/ZAPD ${CMAKE_BINARY_DIR}/ZAPD)
>  add_subdirectory(ZAPDTR/ZAPDUtils ${CMAKE_BINARY_DIR}/ZAPDUtils)
>  add_subdirectory(OTRExporter)
>  add_subdirectory(soh)
> -if(NOT CMAKE_SYSTEM_NAME MATCHES "Darwin|NintendoSwitch|CafeOS")
> +if(NOT CMAKE_SYSTEM_NAME MATCHES "Darwin|NintendoSwitch|CafeOS|Linux")
>  add_subdirectory(OTRGui)
>  endif()
> ```
> 
> **UPDATE**: Re-added the commit for now. Not sure yet where all the options should live, so not going to tangle all that logic up yet. Make sure to also set `BUILD_OTR_GUI=OFF` again.



> Rebased onto current upstream again, includes the lowered CMake version now, so no need to cherry-pick.
> 
> Turns out I didn't read that commit I linked earlier closely enough, it still tries to source all the dependencies for OTRGui, even if you aren't building it. Will fix that when I have some more time, will most likely tie it to the `USE_X11` option, or skip it entirely if we don't have X11.
> 
> Here's a temporary fix if you want to work around it:
> 
> ```diff
> diff --git a/CMakeLists.txt b/CMakeLists.txt
> index f42fb69c..c3cf566f 100644
> --- a/CMakeLists.txt
> +++ b/CMakeLists.txt
> @@ -122,7 +122,7 @@ add_subdirectory(ZAPDTR/ZAPD ${CMAKE_BINARY_DIR}/ZAPD)
>  add_subdirectory(ZAPDTR/ZAPDUtils ${CMAKE_BINARY_DIR}/ZAPDUtils)
>  add_subdirectory(OTRExporter)
>  add_subdirectory(soh)
> -if(NOT CMAKE_SYSTEM_NAME MATCHES "Darwin|NintendoSwitch|CafeOS")
> +if(NOT CMAKE_SYSTEM_NAME MATCHES "Darwin|NintendoSwitch|CafeOS|Linux")
>  add_subdirectory(OTRGui)
>  endif()
> ```
> 
> **UPDATE**: Re-added the commit for now. Not sure yet where all the options should live, so not going to tangle all that logic up yet. Make sure to also set `BUILD_OTR_GUI=OFF` again.

It fails to build. Just tested on a rpiOS 32bit with cmake 3.18 
Tested with cmake 3.24 and same error.

Zaxus125 commented:
Sorry catching up, have been a tad busy with actual work.

> for libglew-dev I installed that and for sdl2 I think Retropie installs a version of sdl2 based on this:- https://github.com/RetroPie/RetroPie-Setup/blob/master/scriptmodules/supplementary/sdl2.sh Is that version not going to work?

I'll take a look into this in more depth, but it appears that installs `libsdl2-dev`, which probably should be sufficient? I'll try replicating this weekend and taking a look, but it likely also needs to include the cmake info for SDL2. I had a similar issue with this for a different build environment, I ended up compiling my own SDL and pointing cmake to that.

I'm also working on a (hopefully) reproducible cross-compilation toolchain to theoretically make cross-compiling for weird platforms easier via buildroot. Will post that when/if I get the motivation to keep wrangling the makefile.

> It fails to build. Just tested on a rpiOS 32bit with cmake 3.18
> Tested with cmake 3.24 and same error.

Sorry, I'm slightly lost -- what error are you seeing, the same with OTRGui still needing X11 or something else?

foxhound311 commented:
> Sorry catching up, have been a tad busy with actual work.
> 
> > for libglew-dev I installed that and for sdl2 I think Retropie installs a version of sdl2 based on this:- https://github.com/RetroPie/RetroPie-Setup/blob/master/scriptmodules/supplementary/sdl2.sh Is that version not going to work?
> 
> I'll take a look into this in more depth, but it appears that installs `libsdl2-dev`, which probably should be sufficient? I'll try replicating this weekend and taking a look, but it likely also needs to include the cmake info for SDL2. I had a similar issue with this for a different build environment, I ended up compiling my own SDL and pointing cmake to that.
> 
> I'm also working on a (hopefully) reproducible cross-compilation toolchain to theoretically make cross-compiling for weird platforms easier via buildroot. Will post that when/if I get the motivation to keep wrangling the makefile.
> 
> > It fails to build. Just tested on a rpiOS 32bit with cmake 3.18
> > Tested with cmake 3.24 and same error.
> 
> Sorry, I'm slightly lost -- what error are you seeing, the same with OTRGui still needing X11 or something else?

i get it to build but it crashes upon start up. If I revert the commit awe44 pointed out it errors out compiling. Not sure what the error is. I do believe awe44 got it running on 64bit tho.

Zaxus125 commented:
> i get it to build but it crashes upon start up. If I revert the commit awe44 pointed out it errors out compiling. Not sure what the error is. I do believe awe44 got it running on 64bit tho.

Oh right, the 32-bit issue, this is also on my to-do list. When I get the time, I'll try to replicate this as well and see if it is something I can fix.

awe444 commented:
Regarding the reverted commit not fixing the 32-bit builds— I think what’s happened is that there’s been more recent commits that the revert doesn’t cleanly merge with. Meaning either we need to revert additional newer commits, or find and fix the root cause. But 64-bit is succeeding as-is, you just might need the extra 1-2GB of swap to accommodate the build.

awe444 commented:
Hi @Zaxus125, I've been regularly building and testing this PR after rebasing it off latest upstream `develop` branch, luckily without any real merge conflict issues, up until yesterday when [this commit](https://github.com/HarbourMasters/Shipwright/commit/11b6ce819d225ea28aa0c0612409e2b5eb75a1d7) was made to CMakeLists.txt and it finally had a conflict I couldn't resolve. Will happily just rebase off a fixed prior commit until you have time to look at it, but just wanted to bring to your attention.

dcvz commented:
Hi @Zaxus125 sorry for getting to this quite late. I think the additions are sound ones. Would you bind breaking it up into separate PR's for the areas they touch? Then the feedback can be for individual parts.

Zaxus125 commented:
Hi, terribly sorry for the long delay, have been busy with actual work and my ever fickle attention span :)

I need to update this branch anyway since it's probably very behind, so it's no extra trouble to open a PR for each option provided they are relatively independent.

Zaxus125 commented:
I have opened up three PRs each containing one component of this PR as completely separate and independent changes. I don't know which changes might be more controversial, so I figured keeping them independent was safer for now as they might be merged in a different order. If/when one gets merged, I will rebase/update each of the rest.

Unfortunately, the OpenGL/GLEW/GLX are somewhat tightly bound together, so I have opted to only open two of those for now, rather than attempt to detangle them all right now.

---

`NO_PULSE` - #1922
`USE_OPENGLES` - #1924
`NO_GLEW` - #1925

vanfanel commented:
@Zaxus125 This is really important to me: I really thank you for fixing the dependencies of this engine. I hope to see those merged soon :)

briaguya0 commented:
i'm going to close this now that it's been split up (and now that most of it needs to be PR'd upstream to LUS)

thanks for your hard work on this! looking forward to seeing it come together with the new architecture!
